src/osyris/core/tools.py

```diff
@@ -38,9 +39,8 @@ def to_bin_edges(x):
     Convert array centers to edges
     """
     centers = to_bin_centers(x)
-    left = centers[0] - (x[1] - x[0])
-    right = centers[-1] + (x[-1] - x[-2])
-    return np.concatenate(np.concatenate(left, centers), right)
+    dx = x[1] - x[0]
+    return np.append(centers - dx, np.array([centers[-1], centers[-1] + dx]))
```

Comment:
While I can see that there is an issue with my implementation (it actually does not work, it throws an error), I do not agree with this change.
Your implementation assumes that all the cells have the same width, which is not necessarily the case.

I suggest to replace with
```Py
def to_bin_edges(x):
    centers = to_bin_centers(x)
    left = centers[0] - (x[1] - x[0])
    right = centers[-1] + (x[-1] - x[-2])
    return np.append(np.insert(centers, 0, left), right)
```

Consider
```Py
a = np.arange(11.)
to_bin_edges(a)
```
Both your implementation and the suggested one above would return
```
[-0.5  0.5  1.5  2.5  3.5  4.5  5.5  6.5  7.5  8.5  9.5 10.5]
```
as expected. However, for
```Py
b = np.arange(11.)**2
to_bin_edges(b)
```
your implementation returns
```
[-0.5  1.5  5.5 11.5 19.5 29.5 41.5 55.5 71.5 89.5 90.5 91.5]
```
while what I am suggesting yields
```
[ -0.5   0.5   2.5   6.5  12.5  20.5  30.5  42.5  56.5  72.5  90.5 109.5]
```
In the figure below, you can see the `edges1` (my suggestion) in orange are a better representation of bin edges
![Screenshot at 2021-11-09 23-14-29](https://user-images.githubusercontent.com/39047984/141013606-9f885c84-becc-4ba5-8c21-72bcacf81806.png)

Reply:
I hadn't thought of that. Nice.